### PR TITLE
Skip logging telemetry on liveness, readiness, health routes.

### PR DESCRIPTION
**Pull request by dtrejod**
## Before this PR
The `/readiness`, `/liveness` , and `/health` endpoints are hit a high frequency producing a high volume of low signal logs. 

## After this PR
By default, disable all request, trace and metric logging on the `/readiness`, `liveness`, and `/health` endpoint by using a new SkipTelemetry RouteParam. 
==COMMIT_MSG==
Skip logging telemetry on liveness and readiness routes.
==COMMIT_MSG==

## Possible downsides?
We no longer will have server side logs for these management endpoints. However we will still have logs on the client side for tracking non 2xx responses.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/257)

### CODE REVIEW

**Review by andybradshaw — COMMENTED** [review]
> Reviewed 9 of 9 files at r1.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/257)* status: all files reviewed, 3 unresolved discussions (waiting on @bmoylan and @dtrejod)

___
*[status/routes/routes.go, line 37 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_fLU-NeqHf2SE5L1:-MUy_fLU-NeqHf2SE5L2:btx5u2p) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/cf4e5b760aeb9eac8846fe1de5a307d1258aafb9/status/routes/routes.go#L37)):*
> ```Go
> 
> func AddHealthRoutes(resource wresource.Resource, source healthstatus.HealthCheckSource, sharedSecret refreshable.String, healthStatusChangeHandlers []status.HealthStatusChangeHandler) error {
> 	return resource.Get("health", status.HealthEndpoint, status.NewHealthCheckHandler(source, sharedSecret, healthStatusChangeHandlers))
> ```

Think we want to skip telemetry here as well? 
___
*[wrouter/route_params.go, line 21 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUyaQQD-v6mFEtR9AHo:-MUyaQQD-v6mFEtR9AHp:b-cndshc) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/cf4e5b760aeb9eac8846fe1de5a307d1258aafb9/wrouter/route_params.go#L21)):*
> ```Go
> )
> 
> type RouteParam interface {
> ```

Not sure how other people feel about this, but it seems a little awkward to be using the interface this way... if we pared this down to a single function interface and used a type switch in each of the `to*` functions, it could remove a lot of needless code:

```golang
type RouteParam interface {
    isRouteParam()
}

type SkipTelemetryParam struct {
    skip bool
}

// SkipTelemetry is a RouterParam that will disable telemetry logging for a route.
func SkipTelemetry(skip bool) RouteParam {
    return SkipTelemetryParam{skip: skip}
}

func (s SkipTelemetryParam) isRouteParam() {}

func toSkipTelemetry(params []RouteParam) bool {
    for _, param := range params {
        switch param.(type) {
        case SkipTelemetryParam:
            return true
        default:
        }
    }
    return false
}
```

It doesn't make sense to me that all types satisfying the interface should have to know about the functionality of the other types satisfying the interface.
___
*[wrouter/router_root.go, line 144 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUygiqtA4AX9aGg8pqK:-MUygiqtA4AX9aGg8pqL:bqevtja) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/cf4e5b760aeb9eac8846fe1de5a307d1258aafb9/wrouter/router_root.go#L144)):*
> ```Go
> 			ParamPerms:    toRequestParamPerms(params),
> 			MetricTags:    metricTags,
> 			SkipTelemetry: toSkipTelemetry(params),
> ```

nit: similar to metrics, I believe this only needs to get evaluated when we register, not inside the handler. Same with the PathPerms I believe.

**Review by bmoylan — COMMENTED** [review]
> Reviewed 2 of 9 files at r1, 1 of 7 files at r3, 6 of 6 files at r4.
> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/257)* status: all files reviewed, 4 unresolved discussions (waiting on @dtrejod and @nmiyake)

___
*[status/routes/routes.go, line 37 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_fLU-NeqHf2SE5L1:-MVHgGXE9oA-b5D7wAIm:b-d4v6fv) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/cf4e5b760aeb9eac8846fe1de5a307d1258aafb9/status/routes/routes.go#L37)):*
<details><summary><i>Previously, andybradshaw (Andy Bradshaw) wrote…</i></summary><blockquote>

Think we want to skip telemetry here as well? 
</blockquote></details>

yep, thanks!
___
*[wrouter/route_params.go, line 105 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MVHgNAE49GwiZ9ItDjz:-MVHgNAE49GwiZ9ItDk-:b3u6pls) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/wrouter/route_params.go#L105)):*
> ```Go
> 
> // SkipTelemetry is a RouterParam that will disable telemetry logging for a route.
> func SkipTelemetry(skip bool) RouteParam {
> ```

Do we expect to ever call this with `false`? Should we just remove the argument?
___
*[wrouter/router.go, line 84 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MVHgiA8CoPc5hog42B3:-MVHgiA8CoPc5hog42B4:b-qjtv9q) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/wrouter/router.go#L84)):*
> ```Go
> 	ParamPerms    RouteParamPerms
> 	MetricTags    metrics.Tags
> 	SkipTelemetry bool
> ```

maybe this should get a comment explaining what it means

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,4 @@
 /out/
 .DS_Store
 /witchcraft/var/
+*.swp
```
Comment (nmiyake) [inline .gitignore]: This should be done as a separate commit/PR

File: changelog/@unreleased/pr-257.v2.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+type: improvement
+improvement:
+  description: Skip logging telemetry on liveness, readiness, and health routes.
```
Comment (nmiyake) [inline changelog/@unreleased/pr-257.v2.yml]: Can you update the changelog description to make it more clear what is being changed? Just reading this in the release notes wouldn't be sufficient for me to fully understand the change.

What telemetry is being skipped -- request logs, service logs, something else? Is it being disabled by default or removed completely? What is the motivation of the change? Having some of this context will greatly help understanding the change later.

File: wrouter/router.go [hunk 3]
```diff
@@ -81,6 +81,7 @@ type RequestVals struct {
 	PathParamVals map[string]string
 	ParamPerms    RouteParamPerms
 	MetricTags    metrics.Tags
+	SkipTelemetry bool
```
Comment (nmiyake) [inline wrouter/router.go]: Wan we use "Disable" rather than "Skip" for the naming?

File: wrouter/route_params.go [hunk 4]
```diff
@@ -18,120 +18,107 @@ import (
...
+		return b.paramPerms
+	}
+	return &requestParamPermsImpl{}
 }
 
-func toMetricTags(params []RouteParam) metrics.Tags {
-	var tags metrics.Tags
-	for _, param := range params {
-		tags = append(tags, param.metricTags()...)
+func toMetricTags(b *routeParamBuilder) metrics.Tags {
```
Comment (nmiyake) [inline wrouter/route_params.go]: What's the reasoning/motivation behind this change?

File: wrouter/route_params.go [hunk 5]
```diff
@@ -18,120 +18,107 @@ import (
...
+	metricTags    metrics.Tags
+	skipTelemetry bool
 }
 
-type requestParamPermsFunc func() RouteParamPerms
-
-func (f requestParamPermsFunc) perms() RouteParamPerms {
-	return f()
+type RouteParam interface {
+	apply(*routeParamBuilder) error
```
Comment (nmiyake) [inline wrouter/route_params.go]: Can you explain why this refactor is being performed?

**Review by nmiyake — CHANGES_REQUESTED** [review]
Thanks for the PR! Understand and agree with the high-level goal and don't see any major flags with the implementation. However, would like to clean this up a bit from a process perspective.

Can you please separate the change that adds a parameter that disables telemetry from the portion that actually makes use of that change? Seeing how all of it works together in this PR is helpful, but would like to evaluate and merge each portion separately.

Also, can you clarify which of the refactors that are included here are needed to implement the new feature versus what is just refactoring for the sake of cleaning up/simplification? This will help the legibility of the change/PR when looking back on it in the future.

**Review by dtrejod — COMMENTED** [review]
> Can you please separate the change that adds a parameter that disables telemetry from the portion that actually makes use of that change? Seeing how all of it works together in this PR is helpful, but would like to evaluate and merge each portion separately.

I created a separate PR with the RouteParam changes so we can keep discussion on the RouteParam refactor there:
https://github.com/palantir/witchcraft-go-server/pull/259

We'll have to merge this PR before moving forward with this PR.

> Also, can you clarify which of the refactors that are included here are needed to implement the new feature versus what is just refactoring for the sake of cleaning up/simplification?

Yes, once the above PR is merged the scope of this PR will be more streamlined. 

> *[Reviewable](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-:-MVMujil6yuB07cTtJm4:b-9ugnu)* status: all files reviewed, 8 unresolved discussions (waiting on @andybradshaw, @bmoylan, and @nmiyake)

___
*[.gitignore, line 8 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_DFhjS_AqWE2h7ir-r4-8:-MVMvGmF0EXynBPrc2rL:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/.gitignore#L8)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

This should be done as a separate commit/PR
</blockquote></details>

Done.
___
*[changelog/@unreleased/pr-257.v2.yml, line 3 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_Duzi7bI6K3fsnVX-r4-3:-MVN0AMkAHnb3TMNXbGZ:b-ljlm6o) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/changelog/@unreleased/pr-257.v2.yml#L3)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

Can you update the changelog description to make it more clear what is being changed? Just reading this in the release notes wouldn't be sufficient for me to fully understand the change.

What telemetry is being skipped -- request logs, service logs, something else? Is it being disabled by default or removed completely? What is the motivation of the change? Having some of this context will greatly help understanding the change later.
</blockquote></details>

Changelog has been updated.
___
*[status/routes/routes.go, line 37 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_fLU-NeqHf2SE5L1:-MVHeq1-Cfc6mXO0rhF5:bb7f2mf) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/cf4e5b760aeb9eac8846fe1de5a307d1258aafb9/status/routes/routes.go#L37)):*
<details><summary><i>Previously, andybradshaw (Andy Bradshaw) wrote…</i></summary><blockquote>

Think we want to skip telemetry here as well? 
</blockquote></details>

done
___
*[wrouter/route_params.go, line 21 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUyaQQD-v6mFEtR9AHo:-MVHerlm4ayS44OHyxcp:byzt2x6) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/cf4e5b760aeb9eac8846fe1de5a307d1258aafb9/wrouter/route_params.go#L21)):*
<details><summary><i>Previously, andybradshaw (Andy Bradshaw) wrote…</i></summary><blockquote>

Not sure how other people feel about this, but it seems a little awkward to be using the interface this way... if we pared this down to a single function interface and used a type switch in each of the `to*` functions, it could remove a lot of needless code:

```golang
type RouteParam interface {
    isRouteParam()
}

type SkipTelemetryParam struct {
    skip bool
}

// SkipTelemetry is a RouterParam that will disable telemetry logging for a route.
func SkipTelemetry(skip bool) RouteParam {
    return SkipTelemetryParam{skip: skip}
}

func (s SkipTelemetryParam) isRouteParam() {}

func toSkipTelemetry(params []RouteParam) bool {
    for _, param := range params {
        switch param.(type) {
        case SkipTelemetryParam:
            return true
        default:
        }
    }
    return false
}
```

It doesn't make sense to me that all types satisfying the interface should have to know about the functionality of the other types satisfying the interface.
</blockquote></details>

Yes, agreed this interface is awkward. I refactored the RouteParam interface require a build helper type that cleans this up. Let me know what you think of this approach. 
___
*[wrouter/route_params.go, line 28 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_DFhjS_AqWE2h7iw-r4-28:-MVMveA44Yo7Oqsg1Ou-:bsxhu65) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/wrouter/route_params.go#L28)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

Can you explain why this refactor is being performed?
</blockquote></details>

The interface for adding a new option to RouteParam was awkward before so I refactor this bit to use builder pattern we use in other projects like `conjure-go-runtime` httpclient params.  This will allow us to continue adding more params in the future without having to extend this RouteParam interface with every new option we want to add. 
___
*[wrouter/route_params.go, line 105 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MVHgNAE49GwiZ9ItDjz:-MVMw7kYDhON2H2oU4bN:b-4fm1yd) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/wrouter/route_params.go#L105)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

Do we expect to ever call this with `false`? Should we just remove the argument?
</blockquote></details>

Good callout. This param should never be called with `false` param since the default is `false`. It also would be confusing which RouteParam would take precedence if one were to specified multiple RouteParams with different SkipTelemetry values. 

Removed param from this method.
___
*[wrouter/route_params.go, line 119 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_DFhjS_AqWE2h7iw-r4-119:-MVMxq084wwzCsCVBakj:b-y6yxvo) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/wrouter/route_params.go#L119)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

What's the reasoning/motivation behind this change?
</blockquote></details>

I put this in the PR description but we want to have a new RouteParam that allows us to disable telemetry for the sole purpose of reducing the number of low signals logs these management endpoints produce. I also put this context for this change in the changelog.
___
*[wrouter/router.go, line 84 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MVHgiA8CoPc5hog42B3:-MVMwYsi4nhiaufqUnMO:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/wrouter/router.go#L84)):*
<details><summary><i>Previously, bmoylan (Brad Moylan) wrote…</i></summary><blockquote>

maybe this should get a comment explaining what it means
</blockquote></details>

Done.
___
*[wrouter/router.go, line 84 at r4](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUy_DFhjS_AqWE2h7ix-r4-84:-MVN061TAOYWLB-M5rV0:b-896fix) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/8c681cc8589d03f5c588add01d67491b05cac311/wrouter/router.go#L84)):*
<details><summary><i>Previously, nmiyake (Nick Miyake) wrote…</i></summary><blockquote>

Wan we use "Disable" rather than "Skip" for the naming?
</blockquote></details>

Done.
___
*[wrouter/router_root.go, line 144 at r1](https://reviewable.io/reviews/palantir/witchcraft-go-server/257#-MUygiqtA4AX9aGg8pqK:-MVHfE8xAzcy3WLdJuTf:bb7f2mf) ([raw file](https://github.com/palantir/witchcraft-go-server/blob/cf4e5b760aeb9eac8846fe1de5a307d1258aafb9/wrouter/router_root.go#L144)):*
<details><summary><i>Previously, andybradshaw (Andy Bradshaw) wrote…</i></summary><blockquote>

nit: similar to metrics, I believe this only needs to get evaluated when we register, not inside the handler. Same with the PathPerms I believe.
</blockquote></details>

done

**Review by bmoylan — APPROVED** [review]
Looks good from my end, but will give @nmiyake a chance to review the update

**Review by nmiyake — APPROVED** [review]
Thanks!

**Pull request merged**